### ISSUE TITLE
Make it that Audio Switcher can use our lib

### ISSUE THREAD

**Issue by vbfox** [AArnott/pinvoke#81]
[Audio Switcher](https://github.com/davkean/audio-switcher) by @davkean is a cool project that use [quite a bit of P/Invoke](https://github.com/davkean/audio-switcher/tree/master/src/AudioSwitcher/Presentation/Drawing/Interop) and it would be nice to have enough API available that we can send it a PR to use our library.

What we miss as APIs :
- [x] `Kernel32\EnumResourceNames`
- [x] `Kernel32\FindResource`
- [x] `Kernel32\LoadResource`
- [x] `Kernel32\LockResource`
- [x] `Kernel32\SizeofResource`
- [x] `Kernel32\LoadLibraryEx`
- [x] `User32\LookupIconIdFromDirectory`
- [x] `User32\LookupIconIdFromDirectoryEx`
- [x] `User32\GetDC`
- [x] `User32\ReleaseDC`
- [x] `Gdi32\GetDeviceCaps`
- [x] `Uxtheme\GetThemeMargins`

Except for the last one it's on dlls we already have in the codebase and the [license of the project](https://github.com/davkean/audio-switcher/blob/master/LICENSE) is also MIT so we can reuse directly the declaration after adaptation.

We also need a few types :
- [x] `GRPICONDIR`
- [x] `GRPICONDIRENTRY`
- [x] `ICONDIR`
- [x] `ICONDIRENTRY`
  Even if the lib is never used in Audio Switcher in the end we will have gained a few more functions.

**Comment by vbfox** [comment]
There are also a few types used while reading resources that aren't directly referenced by `DllImport` declarations but might be included : 
- `GRPICONDIR`
- `GRPICONDIRENTRY`
- `ICONDIR`
- `ICONDIRENTRY`

While they aren't represented directly in the API they represent what `LockResource` would return for an icon so it's quite important to have them to make it usable and I think that it goes into the library scope. What do you thing @AArnott ?

**Comment by AArnott** [comment]
Yes, declaring interop types fits in this project's scope. :)

**Comment by vbfox** [comment]
The last PR should add the missing bits, the remaining action is to prepare a version of audio switcher with all interop replaced by `PInvoke` and check that it still works as expected.

**Comment by vbfox** [comment]
The `audio-switcher` branch is there : https://github.com/vbfox/audio-switcher/tree/pinvoke If it's ok I'll keep this issue open until we have a stable release and can send them a PR.

**Comment by AArnott** [comment]
Sounds good

**Comment by AArnott** [comment]
That's quite the set of changes you've made, @vbfox. The time you took is notable.
I hope davkean likes it. He isn't the type who is afraid of doing pinvoke himself, but perhaps he'll entertain the value of a dedicated P/Invoke library. I know him quite well, but I'm holding out asking him about it now to get his take on it since it's your work.

**Comment by vbfox** [comment]
The changes were mainly mechanical, that's a good news because it mean that we can expect users doing ports to have a relatively painless experience (As long as we have the API they want)

Audio-switcher readme cites http://www.codeproject.com/Articles/32617/Extracting-Icons-from-EXE-DLL-and-Icon-Manipulatio as the source of the Icon code & associated PInvoke so maybe he'll be more open to have them come from an external library rather that include it directly.

I didn't know that you knew each other :)

We'll see how he takes the PR, anyway I selected the project because I use it and it was using old APIs like the resources ones with heavy usage of low level code involving casting `void*` to arbitrary structures, delegates for enumeration that take pointer types, pointers being considered as magic integer values, ...
A lot of that is rarely done from the safety of the managed world and being able to do it while following the code conventions we set for ourselves prove that the library design is sound. So whenever the PR is accepted or not it shown that our current design can accommodate some of the most low level parts of the windows API & we gained a few more functions.
The PR being accepted would only be icing on the cake.
